cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -662,7 +684,7 @@ private boolean rebalanceByMovingLoadOut(Broker broker,
       Broker b = maybeApplyBalancingAction(clusterModel, replica, eligibleBrokers, actionType, optimizedGoals);
       // Only check if we successfully moved something.
       if (b != null) {
-        if (isLoadUnderBalanceUpperLimit(broker)) {
+        if (isLoadUnderBalanceUpperLimit(broker) && !(_fixOfflineReplicasOnly && !broker.currentOfflineReplicas().isEmpty())) {
```

Comment:
If the broker is still over the balance upper limit, and if it can push it under this limit by moving immigrant replicas to the other brokers, this condition enable it to achieve that.
Though, this goal seems to consider all replicas rather than just the offline and immigrant replicas during self healing. I believe we can discuss such behavior changes in another patch.